### ISSUE TITLE
Allow program to continue if stdout is broken

### ISSUE THREAD

**Issue by ctheune** [hynek/structlog#132]
I have a situation where we may encounter a long-running process that is owned by some tool (Consul in this case). We have seen issues where Consul may go away while the sub-process is humming away (like with a live migration). The process continues but the stdout will be broken. At the moment this results in:

```
2017-06-09T23:10:05.726308Z 3973 D blackbee72 migration-failed               action='rescue'
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/fc/qemu/outgoing.py", line 25, in __call__
    self.migrate()
  File "/usr/lib64/python2.7/site-packages/fc/qemu/outgoing.py", line 106, in migrate
    output=pprint.pformat(stat))
  File "/usr/lib64/python2.7/site-packages/structlog/_base.py", line 177, in _proxy_to_logger
    return getattr(self._logger, method_name)(*args, **kw)
  File "/usr/lib64/python2.7/site-packages/structlog/_loggers.py", line 76, in msg
    until_not_interrupted(self._flush)
  File "/usr/lib64/python2.7/site-packages/structlog/_utils.py", line 24, in until_not_interrupted
    return f(*args, **kw)
IOError: [Errno 32] Broken pipe
```

I think we would be fine if the PrintLogger would be able to ignore (maybe not by default) if stdout has vanished.

Our specific code involved that I have seen failing is this:
https://bitbucket.org/flyingcircus/fc.qemu/src/96134201bdc27cf8448dcc5e2531fb94aa2226cb/src/fc/qemu/outgoing.py?at=default&fileviewer=file-view-default#outgoing.py-102

Thoughts?

**Comment by hynek** [comment]
Uh how do other logging systems handle that?  Silently dropping log entries sound slightly evil to me. :)

P.S. 2.7? Tsk.

**Comment by ctheune** [comment]
:P~

Well "it depends". If you log to other targets, i.e. syslog UDP then you don't care.

I mean: do you want errors in the logging system to propagate in a disastrous way? This code specifically is quite close to system-level components and I'm fighting the "last" edge cases that only hit us in a blue moon but still - from a reliability perspective ... 

Additionally, my logging is set up with a console logger for the interactive console, but the logging also sends things off to disk in parallel. Which reminds me that that may also fail unexpectedly ... sigh. Wasn't logging supposed to be a source of stability, not fragility? :)

**Comment by hynek** [comment]
So PrintLogger is a handful of lines (https://github.com/hynek/structlog/blob/master/src/structlog/_loggers.py#L39-L79) – I think it would be more prudent to add a DontGiveAFuckPrintLogger that tries to ignore everything instead of adding options.

> Wasn't logging supposed to be a source of stability, not fragility? :)

HAHAHAHAHAHAH

**Comment by ctheune** [comment]
I'm happy with adding a DGAFLogger. I also noticed that my code would become cleaner when doing a DGAFMultiLogger that allows me to manage multiple targets. Probably not the smartest code written using structlog, but hey. I'll pass a link to the code once I pushed it.

Nevertheless, these kinds of edge conditions are quite "interesting" when dealing with system-level stuff that doesn't have "crash early" as an option when facing adverse conditions. :)
